### ISSUE TITLE
SuperDevMode does not start automatically

### ISSUE THREAD

**Issue by dnouls** [gwt-plugins/gwt-eclipse-plugin#332]
I'm using a project structure based on the tbroyer multi-module archetype.
I have the facet project properties configured to launch super dev mode, but I always get the following error in the error log of eclipse.

!ENTRY com.gwtplugins.gwt.eclipse.wtp 1 1 2017-01-05 14:59:31.955
!MESSAGE GWT: posiblyLaunchGwtSuperDevModeCodeServer: GWT Facet project properties, the code server sync is off.

It would be useful if it would tell me which project it is looking at. I have configured the client/server and shared projects to enable the feature but the plugin is apparently looking somewhere else ?

**Comment by dnouls** [comment]
Any feedback ?

**Comment by branflake2267** [comment]
Good point. I've marked it as a bug and will add more logging on the next round. 

If you haven't gone down this route already, what I would suggest doing to narrow it down is close project in your workspace to drill it.

**Comment by branflake2267** [comment]
Things are working smoothly. I'll continue to test, but so far, things seem better. 

![screen shot 2017-03-01 at 2 23 53 pm](https://cloud.githubusercontent.com/assets/1326504/23484303/4c06f752-fe8b-11e6-90b3-73946d6637c3.png)

**Comment by branflake2267** [comment]
Things are working as expected in tests.

**Comment by dnouls** [comment]
I'll do some testing on my own projects tomorrow. thanks.

**Comment by branflake2267** [comment]
Quick note, the are in the staging/snapshot repository and not production yet.

**Comment by branflake2267** [comment]
I need to figure out how to replicate your project so I can test this config. Could you make a project test case I could use for testing?

**Comment by branflake2267** [comment]
I just noticed my snapshots weren't uploading to the staging/snapshot repo. It should be up soon, if all works as expected. So the changes haven't been updated yet.

**Comment by dnouls** [comment]
I will try to create an example project setup that exposes the problem. I need to do this on my computer at home (Enterprise hell)

**Comment by branflake2267** [comment]
Sounds good, thanks.

**Comment by dnouls** [comment]
Trying to reproduce by creating a new project.
1) Create a new multimodule project (using 1.0-SNAPSHOT of tbroyers archetypes)
Started tomcat with the server project.
Error log states that the project (does not say which one) does not have a GWT Facet.
2) Added (manually) the GWT Facet to the server project
Error log still states that the project does not have a GWT Facet
3) Added (manually) the GWT Facet to the client project
Now it all starts automatically.

So first question:
How can I get eclipse to add the GWT Facet automatically when importing the maven project ?
Maybe a fix is required in the GWT maven archetype ?

I will continue now to see what happens when I add gwt-libs in the mix

**Comment by dnouls** [comment]
Added gwt-lib projects... but it keeps on working properly!
So there must be something very specific to my pom files that causes my problem.
Not sure how to progress, I cannot share my full application workspace with you.
I guess you can close this issue until I can reproduce and proof that it is a problem in the plugin or my own pom mistakes.

**Comment by dnouls** [comment]
Aha, I'm reaching some conclusions now.
I have multiple gwt-libs that have the GWT Facet enabled in the eclipse projects.
It looks like the GWT plugin is assuming the very first project to contain the entry point for the war.

**Comment by branflake2267** [comment]
If the facet isn't getting wired up on import of the gwt project, that's a maven configurator issue in the plugin. The GWT plugin has to be available in the pom.xml for it to be turned on.

**Comment by branflake2267** [comment]
I'll test importing the archetypes on my next round of fixes and verify that the configurator is adding the facet. Thanks for the report.

**Comment by dnouls** [comment]
the maven plugin is in the pom, so I don't think I am doing anything wrong there.
Anyway, I'm soo happy that it works now, it makes debugging and launching soo much easier.
Great work, thanks.

**Comment by branflake2267** [comment]
Cool, good to hear the debugging. I've noticed at times things don't autowire and tried to solve those on the last fixes. I'll keep this open and do some more investigation to see if I can replicate it. Thanks for the feedback too!

**Comment by dnouls** [comment]
After doing a maven project update the GWT facets are back on my gwt-lib projects and as a consequence the auto launch feature does not work anymore. Where is the code in the plugin that decides which eclipse project should be used to launch super dev mode ? It does not make sense to select a gwt-lib project, it only makes sense on a gwt-app project.

**Comment by branflake2267** [comment]
Gotcha, thanks for showing me that. So I need to setup a test case so I can work that out.

**Comment by xak2000** [comment]
Hi!
I found another issue with `GWT: possiblyLaunchGwtSuperDevModeCodeServer: Does not have a GWT Facet.`

My project structure is simple:
```
|- main-project-module
|--- gwt-module (client & server)
|--- utils
|--- persistence
```

When I deploy the `gwt-module` to Eclipse's Tomcat server, I publish only one modle `gwt-module`. The other moduels aren't web-apps and can't be published at all, it is ok. They are just dependencies of `gwt-module`. The `persistence` module even has gwt module descriptor to be able to use some it's classes from `gwt-module`'s client side.
But from the Tomcat server perspective the published `gwt-module` module has submodules: `utils`, `persistence`. So when published, it looks like:
```
Tomcat 8.0 server at localhost
|- gwt-module
|--- utils
|--- persistence
```
To be more clear, there is screenshot:
![Screenshot](http://dl4.joxi.net/drive/2017/04/27/0018/1284/1230084/84/76b3fa8c78.jpg)

The only module which has GWT facet is `gwt-module` (`point-manager` on screenshot).
The `Sync` flag is ON.
But I still get the error `GWT: possiblyLaunchGwtSuperDevModeCodeServer: Does not have a GWT Facet.` when I start Tomcat.

I really don't know why this happens, but have a guess:
I looked at [GwtFacetUtils.getModules(...)](https://github.com/gwt-plugins/gwt-eclipse-plugin/blob/33a5f066d6efe3bd519fb8884de4f786ace82550/plugins/com.gwtplugins.gwt.eclipse.wtp/src/com/google/gwt/eclipse/wtp/utils/GwtFacetUtils.java#L157) function and I think it can't find my main gwt module because it looks at child modules first, and, if any child modules found, it returns just these child modules, ignoring parent module:

```
    // Multi-module project
    IModule[] modules = server.getChildModules(server.getModules(), new NullProgressMonitor());
    if (modules == null || modules.length == 0) { // does it have multi-modules?
      // So it doesn't have multi-modules, lets use the root as the module.
      modules = server.getModules();
    }
```

It return parent modules only if child modules are not found.

But in my project it is parent module is actually has GWT facet. It is parent module should be lauched in CodeServer.

I think the simple solution is to modify `getModules` method to return the full list of modules at any level of hierarchy, recursively doing `server.getChildModules` and including all found modules (parent and child) in returned array.
